### ISSUE TITLE
Compile error in traits.hpp on OSX

### ISSUE THREAD

**Issue by astrodroid** [USCiLab/cereal#341]
Hi,
 
I am using cereal to load game scenario/map data from file. So far this was working very well. But a couple of days ago, I had to move the classes that cereal serialises away from the game project in a separate static library that is used to calculate the objects position etc..  

However as soon as I reference the cereal headers within that static library I encounter a compile error on the traits.hpp file:
```cpp
    namespace detail
    {
      struct shared_from_this_wrapper
      {
        template <class U>
        static auto **check**( U const & t ) -> decltype( ::cereal::access::shared_from_this(t), std::true_type() );  _//error: Expected member name or ';' after declaration specifiers_

        static auto **check**( ... ) -> decltype( std::false_type() ); _//error: Expected member name or ';' after declaration specifiers_

        template <class U>
        static auto get( U const & t ) -> decltype( t.shared_from_this() );
      };
    }

    //! Determine if T or any base class of T has inherited from std::enable_shared_from_this
    template<class T>
    struct has_shared_from_this : decltype(detail::shared_from_this_wrapper::**check**(std::declval<T>())) _//error: Expected unqualified-id_
    { };
```

The problem is a bit confusing as this only happens when compiling for OSX and not for iOS. If I understood what is happening correctly, it is not that the static check that is failing, but the compiler appears to be calling a macro called check contained within "AssertMacros.h" :

```cpp
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/
MacOSX10.11.sdk/usr/include/AssertMacros.h:1291:28: note: expanded from macro 'check'
        #define check(assertion)  __Check(assertion)
```

At the same time the same piece of code that doesn't work within the staticlibrary works fine if called from the main game project, for both iOS and OSX. 

For now I have worked-around this issue by renaming the check methods within traits.hpp to "check_" which resolves the issue for me.

This is the full error log:
http://pastebin.com/6GPH6D50

**Comment by AzothAmmo** [comment]
See #273 for the same issue. I previously listed this as a nofix but might reconsider if this is going to continue to affect people.

**Comment by astrodroid** [comment]
Ok thanks. 

Yes, setting ```#define __ASSERT_MACROS_DEFINE_VERSIONS_WITHOUT_UNDERSCORES 0 ``` in the prefix folder  prevents the error from occurring.

Really odd naming choice for that C macro.

**Comment by erichkeane** [comment]
@AzothAmmo : It seems that the XCode "check" macro is a function call macro, so only "check(" will cause a problem.  It seems to me that we could very easily avert this issue by renaming the shared_from_this_wrapper functions to something else, even check_(...).  Alternatively, we could change the usages of it to be surrounded with '('')', like we do with min/max (thanks to another stupid macro).

Would you accept either of those patches Shane?  @astrodroid, if I were to put said patch together, could you validate it on OSX?

**Comment by AzothAmmo** [comment]
Either of those solutions is fine. Alternatively check could be renamed test, which would also work unless there's another macro out there with that name too.

**Comment by AzothAmmo** [comment]
merged #342
